api_test_utils/api_session_client.py

```diff
@@ -28,30 +29,121 @@ def _full_url(self, url: StrOrURL) -> StrOrURL:
         url = os.path.join(self.base_uri, url)
         return url
 
-    def get(self, url: StrOrURL, *, allow_redirects: bool = True, **kwargs: Any) -> "aoihttp._RequestContextManager":
+    def _request(
+        self,
+        method: str,
+        url: StrOrURL,
+        *,
+        allow_retries: bool = False,
+        max_retries: int = 5,
+        allow_redirects: bool = True,
+        **kwargs: Any
+    ) -> "aoihttp._RequestContextManager":
         uri = self._full_url(url)
-        return self.session.get(uri, allow_redirects=allow_redirects, **kwargs)
+        if allow_retries:
+            resp = self._retry_requests(
+                lambda: self.session.request(
+                    method, uri, allow_redirects=allow_redirects, **kwargs
+                ),
+                max_retries=max_retries
+            )
+        else:
+            resp = self.session.request(
+                method, uri, allow_redirects=allow_redirects, **kwargs
+            )
+        return resp
 
-    def post(self, url: StrOrURL, *, allow_redirects: bool = True, **kwargs: Any) -> "aoihttp._RequestContextManager":
-        uri = self._full_url(url)
-        return self.session.post(uri, allow_redirects=allow_redirects, **kwargs)
+    async def _retry_requests(self, make_request, max_retries):
+        retry_codes = {429, 503}
+        for retry_number in range(max_retries):
+            resp = await make_request()
+            if resp.status in retry_codes:
+                time.sleep(2**retry_number - 1)
+                continue
+            return resp
+        else:
+            raise TimeoutError("Maxium retry limit hit.")
 
-    def put(self, url: StrOrURL, *, allow_redirects: bool = True, **kwargs: Any) -> "aoihttp._RequestContextManager":
-        uri = self._full_url(url)
-        return self.session.put(uri, allow_redirects=allow_redirects, **kwargs)
+    def get(
```

Comment:
The methods below will need some refactor in cut #2 I think, there's a lot of the same stuff going on in each and only the method param is really changing.